draw/DrawingProgram.py

```diff
@@ -0,0 +1,77 @@
+
+
+class DrawingProgram(ShapeFactory):
+    # For the intents and purposes of this assignment, all shapes should be created via ShapeFactory.
+    #     Thus any adding of shapes to your DrawingProgram class object must be done through ShapeFactory.
+    #     This may seem cumbersome, but the point is to leave object creation to a single class.
+    #         It encapsulates how to create the objects you need.
+    #  ^^^ I think this means we pass in ShapeFactory into DrawingProgram instead of the shapes themselves?
+
+    def __init__(self, shapes = None):
+        if shapes is None or isinstance(shapes, list):
+            self.shapes = shapes
+        else:
+            Exception(f"Shapes: {shapes} is not None or a list.")
+
+    def add_shape(self, shape):
+        self.shapes.append(shape)
+
+    def remove_shape(self, shape):
+        """remove shape that exists in the DrawingProgram.Shapes list, else throw an exception"""
+        shape_rm_count = self.shapes.count(shape)
+        if shape in self.shapes:
+            self.Shapes.remove(shape)
+        else:
+            raise Exception(f'{shape} not in {self.shapes}')
+
+        print(f"{shape_rm_count} number of occurrences removed from shape list.")
+        return shape_rm_count
+
+    def print_shapes(self, shape):
+        """prints all shapes that match the type of the shape passed in.
+        throw exception if shape is not in ShapeFactory shapes."""
+        if shape not in ShapeFactory.shape_dict():
+            Exception(f'{shape} is not a shape in ShapeFactory.')
+
+        matched_shapes = []
+        for each_shape in self.shapes:
+            if isinstance(shape, each_shape):
+                matched_shapes.append(each_shape)
+
+        return matched_shapes
+
+    def sort_shapes(self):
+        """sorts the list/collection of shapes -- you must use a sort
+        that runs in O(nlogn) for its worst case. shapes will be sorted
+        first by name, then by area if names are same"""
+        # Sean's question: can we use preprogrammed sort_shapes(): "sorts the list/collection of shapes --
+        #  ^^ concat name and area and put together to do it
+
+    def __str__(self):
+        """returns a string representation of each of the shapes --
+        each shape will be separated from others by a newline (\n)"""
+
+    def get_shape(self, index):
+        """returns the shape at the specified index"""
+
+    def set_shape(self, index, shape):
+        """replaces the shape at the specified index any other
+        behaviors you feel are necessary for the class"""
+
+
+class DrawingProgramIterator:
+    """DrawingProgramIterator provides the ability to iterate
+    across the collection of shapes in DrawingProgram using a for loop."""
+    # See notes/slides on Iterator and
+    # book chapter "The Iterator Pattern" and the section "The Iterator Protocol"
+    # Sean's question: (why not build it into the class?)
+
+
+class DrawingProgramMain(DrawingProgram):
```

Comment:
I'm going to purpose breaking this out of this file. I think this file will get unmanageable if we add this in there too.